File: tests/html/event_customevent_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library EventCustomEventTest;

import 'package:unittest/unittest.dart';
import 'package:unittest/html_config.dart';
import 'dart:html';
import 'dart:js' as js;

class DartPayloadData {
  final dartValue;

  DartPayloadData(this.dartValue);
}

main() {
  useHtmlConfiguration();

  test('custom events', () {
    var provider = new EventStreamProvider<CustomEvent>('foo');
    var el = new DivElement();

    var fired = false;
    provider.forTarget(el).listen((ev) {
      fired = true;
      expect(ev.detail, {'type': 'detail'});
    });

    var ev = new CustomEvent('foo',
        canBubble: false, cancelable: false, detail: {'type': 'detail'});
    el.dispatchEvent(ev);
    expect(fired, isTrue);
  });

  test('custom events from JS', () {
    var scriptContents = '''
      var event = document.createEvent("CustomEvent");
      event.initCustomEvent("js_custom_event", true, true, {type: "detail"});
      window.dispatchEvent(event);
    ''';

    var fired = false;
    window.on['js_custom_event'].listen((ev) {
      fired = true;
      expect(ev.detail, {'type': 'detail'});
    });

    var script = new ScriptElement();
    script.text = scriptContents;
    document.body.append(script);

    expect(fired, isTrue);
  });

  test('custom events to JS', () {
    expect(js.context['gotDartEvent'], isNull);
    var scriptContents = '''
      window.addEventListener('dart_custom_event', function(e) {
        if (e.detail == 'dart_message') {
          e.preventDefault();
          window.gotDartEvent = true;
        }
        window.console.log('here' + e.detail);
      }, false);''';

    document.body.append(new ScriptElement()..text = scriptContents);

    var event = new CustomEvent('dart_custom_event', detail: 'dart_message');
    window.dispatchEvent(event);
    expect(js.context['gotDartEvent'], isTrue);
  });

  test('custom data to Dart', () {
    var data = new DartPayloadData(666);
    var event = new CustomEvent('dart_custom_data_event', detail: data);

    var future = window.on['dart_custom_data_event'].first.then((_) {
      expect(event.detail.dartValue, 666);
    });

    document.body.dispatchEvent(event);
    return future;
  });
}

File: pkg/dev_compiler/test/codegen_expected/lib/html/event_customevent_test.js
dart_library.library('lib/html/event_customevent_test', null, /* Imports */[
  'dart_sdk',
  'unittest'
], function load__event_customevent_test(exports, dart_sdk, unittest) {
  'use strict';
  const core = dart_sdk.core;
  const html = dart_sdk.html;
  const js = dart_sdk.js;
  const async = dart_sdk.async;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const html_config = unittest.html_config;
  const unittest$ = unittest.unittest;
  const src__matcher__expect = unittest.src__matcher__expect;
  const src__matcher__core_matchers = unittest.src__matcher__core_matchers;
  const event_customevent_test = Object.create(null);
  let EventStreamProviderOfCustomEvent = () => (EventStreamProviderOfCustomEvent = dart.constFn(html.EventStreamProvider$(html.CustomEvent)))();
  let CustomEventTovoid = () => (CustomEventTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [html.CustomEvent])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  let dynamicTovoid = () => (dynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic])))();
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let VoidToFuture = () => (VoidToFuture = dart.constFn(dart.definiteFunctionType(async.Future, [])))();
  event_customevent_test.DartPayloadData = class DartPayloadData extends core.Object {
    new(dartValue) {
      this.dartValue = dartValue;
    }
  };
  dart.setSignature(event_customevent_test.DartPayloadData, {
    constructors: () => ({new: dart.definiteFunctionType(event_customevent_test.DartPayloadData, [dart.dynamic])})
  });
  event_customevent_test.main = function() {
    html_config.useHtmlConfiguration();
    unittest$.test('custom events', dart.fn(() => {
      let provider = new (EventStreamProviderOfCustomEvent())('foo');
      let el = html.DivElement.new();
      let fired = false;
      provider.forTarget(el).listen(dart.fn(ev => {
        fired = true;
        src__matcher__expect.expect(ev[dartx.detail], dart.map({type: 'detail'}));
      }, CustomEventTovoid()));
      let ev = html.CustomEvent.new('foo', {canBubble: false, cancelable: false, detail: dart.map({type: 'detail'})});
      el[dartx.dispatchEvent](ev);
      src__matcher__expect.expect(fired, src__matcher__core_matchers.isTrue);
    }, VoidTodynamic()));
    unittest$.test('custom events from JS', dart.fn(() => {
      let scriptContents = '      var event = document.createEvent("CustomEvent");\n      event.initCustomEvent("js_custom_event", true, true, {type: "detail"});\n      window.dispatchEvent(event);\n    ';
      let fired = false;
      html.window[dartx.on].get('js_custom_event').listen(dart.fn(ev => {
        fired = true;
        src__matcher__expect.expect(dart.dload(ev, 'detail'), dart.map({type: 'detail'}));
      }, dynamicTovoid()));
      let script = html.ScriptElement.new();
      script[dartx.text] = scriptContents;
      html.document[dartx.body][dartx.append](script);
      src__matcher__expect.expect(fired, src__matcher__core_matchers.isTrue);
    }, VoidTodynamic()));
    unittest$.test('custom events to JS', dart.fn(() => {
      src__matcher__expect.expect(js.context.get('gotDartEvent'), src__matcher__core_matchers.isNull);
      let scriptContents = '      window.addEventListener(\'dart_custom_event\', function(e) {\n        if (e.detail == \'dart_message\') {\n          e.preventDefault();\n          window.gotDartEvent = true;\n        }\n        window.console.log(\'here\' + e.detail);\n      }, false);';
      html.document[dartx.body][dartx.append]((() => {
        let _ = html.ScriptElement.new();
        _[dartx.text] = scriptContents;
        return _;
      })());
      let event = html.CustomEvent.new('dart_custom_event', {detail: 'dart_message'});
      html.window[dartx.dispatchEvent](event);
      src__matcher__expect.expect(js.context.get('gotDartEvent'), src__matcher__core_matchers.isTrue);
    }, VoidTodynamic()));
    unittest$.test('custom data to Dart', dart.fn(() => {
      let data = new event_customevent_test.DartPayloadData(666);
      let event = html.CustomEvent.new('dart_custom_data_event', {detail: data});
      let future = html.window[dartx.on].get('dart_custom_data_event').first.then(dart.dynamic)(dart.fn(_ => {
        src__matcher__expect.expect(dart.dload(event[dartx.detail], 'dartValue'), 666);
      }, dynamicTodynamic()));
      html.document[dartx.body][dartx.dispatchEvent](event);
      return future;
    }, VoidToFuture()));
  };
  dart.fn(event_customevent_test.main, VoidTodynamic());
  // Exports:
  exports.event_customevent_test = event_customevent_test;
});
